### PR TITLE
[pt][quant] Vectorized qmul and more methods on qint data types

### PR DESCRIPTION
**Pull request by dskhudia**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34376 [pt][quant] Vectorized qmul and more methods on qint data types**

Vectorized implementation of qmul. qmul is now ~16x faster on my development machine. This implementation works for qint8, quint8 and qint32. Also added some commonly used operations, such as multiply operator, requantize operation etc., to qint vector classes for future use.

```
#!/usr/bin/env python

import time
import torch
import torch.nn as nn
torch.set_num_threads(1)
# print(torch.__config__.parallel_info())

A = torch.rand(1, 54, 54, 256)
B = torch.rand(1, 54, 54, 256)

scale = .05
zero_point = 50

for dtype in [torch.quint8, torch.qint8]:

    qA = torch.quantize_per_tensor(A, scale=scale, zero_point=zero_point,
            dtype=dtype)
    qB = torch.quantize_per_tensor(B, scale=scale, zero_point=zero_point,
            dtype=dtype)

    NITER = 1000
    s = time.time()
    for i in range(NITER):
        out = torch.ops.quantized.mul(qA, qB, scale=scale, zero_point=zero_point)
    time_per_iter = (time.time() - s) / NITER

    print('dtype: {} time per iter ms: {:.3f}'.format(dtype, time_per_iter * 1000))
```
### Before
dtype: torch.quint8 time per iter ms: 6.714
dtype: torch.qint8 time per iter ms: 6.780

### After
dtype: torch.quint8 time per iter ms: 0.431
dtype: torch.qint8 time per iter ms: 0.417

### Test
Modified qmul tests to include qint8 and qint32 data types.

ython test/test_quantized.py TestQuantizedOps.test_qmul_relu_same_qparams
ython test/test_quantized.py TestQuantizedOps.test_qmul_relu_different_qparams
ython test/test_quantized.py TestQuantizedOps.test_qmul_broadcast

Differential Revision: [D20308515](https://our.internmc.facebook.com/intern/diff/D20308515/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20308515/)!

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 1]
```diff
@@ -458,6 +458,58 @@ void qadd_kernel(Tensor& out, const Tensor& self, const Tensor& other) {
...
+          scalar_t res =
+              at::requantize_val<scalar_t>(multiplier, zero_point, c);
+          if (ReLUFused) {
+            res.val_ = std::max<scalar_t::underlying>(res.val_, zero_point);
+          }
+          return res;
+        },
+        [&](Vec a, Vec b) -> Vec {
+          Vec::int_vec_return_type a_sub_zp =
+              a.subtract(Vec(static_cast<scalar_t>(self_zero_point)));
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: Can we name this something like `widening_subtract`? I was very confused when I initially read this as to why it was expanding out to multiple vectors

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 2]
```diff
@@ -458,6 +458,58 @@ void qadd_kernel(Tensor& out, const Tensor& self, const Tensor& other) {
...
+          Vec::int_vec_return_type b_sub_zp =
+              b.subtract(Vec(static_cast<scalar_t>(other_zero_point)));
+          Vec::int_vec_return_type c;
+          for (int i = 0; i < Vec::int_num_vecs(); ++i) {
+            c[i] = a_sub_zp[i] * b_sub_zp[i];
+          }
+          Vec rv = Vec::requantize(c, multiplier, zero_point);
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: Same nit about naming here, maybe `requantize_from_int` so folks can quickly distinguish this from quantization from float

**Review by jamesr66a — APPROVED** [review]
Love it! Small nits about naming to make callsites easier to understand, but the patch makes sense to me and the perf improvements are encouraging

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp (hunk 1 above)
Comment (dskhudia) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: It's not widening_subtract for qint32 but I am not opposed to renaming it to widening_subtract. subtract_int32?

**Pull request closed without merging**